### PR TITLE
Copter: support SET_GPS_GLOBAL_ORIGIN message

### PR DESCRIPTION
**Pull request by rmackay9**
This PR resolves an issue that could cause vehicles to fly to extremely high altitudes during autonomous modes including RTL, Auto, Guided in cases where the ground station has sent an inaccurate altitude in a DO_SET_HOME message before the vehicle's EKF origin has been set (i.e. before a good GPS lock).

The following changes are included:

- DO_SET_HOME commands (received either within a COMMAND_LONG message or SET_HOME_POSITION messages) only set the AHRS home, they no longer set the EKF origin.
- SET_GPS_GLOBAL_ORIGIN messages are consumed and used to set the EKF origin.  When the origin is set a GPS_GLOBAL_ORIGIN message is sent to the ground station
- GET_HOME commands return both home and EKF origin to the caller (there is currently no command that allows requesting the origin)
- when home is set (either automatically or when specified by the user) both the home and origin are sent to all mavlink channels
- a new arming check is added that checks that the AHRS's reported position is within 10m of the GPS position.  This check is within the GPS arming checks so the check will not run in cases where the user is not using a GPS (i.e. a manual flight mode or GPS is not being used)

These changes mostly only affect Copter (the AP_Arming check affects all vehicles) but I will extend the changes to at least Rover and Tracker (and Plane if people want) if the peer review is happy with them.

### CODE REVIEW

File: ArduCopter/GCS_Mavlink.cpp [hunk 1]
```diff
@@ -742,6 +742,22 @@ void GCS_MAVLINK_Copter::handleMessage(mavlink_message_t* msg)
...
 
+    case MAVLINK_MSG_ID_SET_GPS_GLOBAL_ORIGIN:
+    {
+        mavlink_set_gps_global_origin_t packet;
+        mavlink_msg_set_gps_global_origin_decode(msg, &packet);
+        // sanity check location
+        if (!check_latlng(packet.latitude, packet.longitude)) {
+            break;
+        }
+        Location ekf_origin;
```
Comment (tridge) [inline ArduCopter/GCS_Mavlink.cpp]: this in unsafe stack init of a structure. Need to zero, or set options

File: ArduPlane/GCS_Mavlink.cpp [hunk 2]
```diff
@@ -895,6 +895,22 @@ void GCS_MAVLINK_Plane::handleMessage(mavlink_message_t* msg)
...
+            break;
+        }
+        Location ekf_origin {};
+        ekf_origin.lat = packet.latitude;
+        ekf_origin.lng = packet.longitude;
+        ekf_origin.alt = packet.altitude / 10;
+        plane.set_ekf_origin(ekf_origin);
+        break;
+    }
+
```
Comment (amilcarlucas) [inline ArduPlane/GCS_Mavlink.cpp]: Can't this code be moved to the GCS_Mavlink library ?

File: ArduCopter/commands.cpp [hunk 3]
```diff
@@ -94,8 +93,39 @@ bool Copter::set_home(const Location& loc, bool lock)
...
+    // return success
+    return true;
+}
+
+// sets ekf_origin if it has not been set.
+//  should only be used when there is no GPS to provide an absolute position
+//  returns true on success
+bool Copter::set_ekf_origin(const Location& loc)
+{
+    // only allow setting origin if motors are disarmed
```
Comment (OXINARF) [inline ArduCopter/commands.cpp]: Why? The origin will be set if GPS fix is acquired while flying.

**Comment by amilcarlucas** [comment]
For this to be completed I guess https://github.com/ArduPilot/ardupilot/pull/6720 should get merged at some point :)

**Comment by OXINARF** [comment]
@amilcarlucas I don't see how that PR is in any way connected to this one...?

**Comment by amilcarlucas** [comment]
That one also does a lot of stuff with origin and home.

**Comment by tridge** [comment]
I'd suggest one more small change to DO_SET_HOME. I think we should make param1=2 mean "set latitude and longitude, but not altitude". 
For apps that want to do "return to my device" I think mostly they will want to only specify a latitude/longitude and not an altitude. To do that now they have to first fetch HOME_POSITION, then convert the altitude units then send back DO_SET_HOME with the fetched altitude. That is racy and also more difficult than it should be. Instead sending a DO_SET_HOME which doesn't set altitude is easier.

File: ArduCopter/commands.cpp (hunk 3 above)
Comment (rmackay9) [reply ArduCopter/commands.cpp]: Yes, I also wondered if we should allow setting the ekf origin while moving.  I can't think of a reason why we can't so I'll remove that restriction.

File: libraries/AP_Arming/AP_Arming.cpp [hunk 4]
```diff
@@ -374,6 +375,19 @@ bool AP_Arming::gps_checks(bool report)
             }
             return false;
         }
+
+        // check AHRS and GPS are within 10m of each other
+        Location gps_loc = ahrs.get_gps().location();
+        Location ahrs_loc;
+        if (ahrs.get_position(ahrs_loc)) {
```
Comment (WickedShell) [inline libraries/AP_Arming/AP_Arming.cpp]: This appears to break the plane usage scenario where you move the aircraft from where it was booted then attempt to arm. If I left the plane at the new location long enough it was okay, however that took far longer then makes sense/would confuse a user in the field. (Confused me at least)

**Pull request merged**